### PR TITLE
Add ruff to github code check, use local deps for pre-commit

### PR DESCRIPTION
**Pull request by collindutter**
~vacation pr pls ignore~
- Add `ruff` to Github linting code check
- Use local dependencies for all `pre-commit` hooks.
- Update order of `pre-commit` hooks.
- Update dev dependencies.
- Ignore some new pyright errors that conflict with behavior we rely on.
- Add `--fix` to `ruff` `pre-commit` hook. This is the source of all the cleaned up imports.
- Replace `black` with `ruff format`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -151,11 +151,11 @@ pytest-clarity = "^1.0.1"
 optional = true
 
 [tool.poetry.group.dev.dependencies]
-ruff = "^0.0.286"
-black = "^23.7.0"
-pyright = "^1.1.355"
-pre-commit = "^3.5.0"
-boto3-stubs = {extras = ["bedrock", "iam", "opensearch", "s3", "sagemaker"], version = "^1.34.18"}
+ruff = "^0.4.4"
+black = "^23.12.1"
```
Comment (dylanholmes) [inline pyproject.toml:155]: Do we still need black?

[Ruff](https://docs.astral.sh/ruff/) claims:
> Drop-in parity with [Flake8](https://docs.astral.sh/ruff/faq/#how-does-ruff-compare-to-flake8), isort, and Black

**Review by dylanholmes — DISMISSED** [review]
We gonna switch to [uv](https://github.com/astral-sh/uv) too? 😛

**Review by andrewfrench — DISMISSED** [review]
:broom: Clean it up!

File: pyproject.toml (hunk 1 above)
Comment (collindutter) [reply pyproject.toml:155]: Oh dang that's new(ish)! Goodbye black 🫡

**Review by andrewfrench — APPROVED** [review]

**Pull request merged**